lib/teleterm/gateway/gateway_test.go

```diff
@@ -91,7 +95,7 @@ func TestNewWithLocalPortStartsListenerOnNewPortIfPortIsFree(t *testing.T) {
 	tcpPortAllocator := gatewaytest.MockTCPPortAllocator{}
 	oldGateway := createAndServeGateway(t, &tcpPortAllocator)
 
-	newGateway, err := NewWithLocalPort(*oldGateway, "12345")
+	newGateway, err := NewWithLocalPort(oldGateway, "12345")
```

Comment:
I'm unfamiliar with this test code - is there any reason you need a specific port? Can we just listen with ":0" and let the OS allocate an available port?

Reply:
`gateway.NewWithLocalPort` is used in Connect when the user wants to change the port number of the proxy. Take a look at the screenshot that I posted above, there's a field for the port number next to the database name field.

Of course, in tests we don't want to listen on a specific port. But I still want to test that when the user asks the app to change the port to 12345, then the port is actually changed to that number. So instead of using `net.Listen` in `gateway.Gateway`, I added `TCPPortAllocator` interface. In tests I use mock listeners which obtain a random port and `mockListener.Addr()` reports a fake address, in this case it'd be `localhost:12345`.

https://github.com/gravitational/teleport/blob/7676f3819937cdaeeaf19aedd57ca3db796c962e/lib/teleterm/gatewaytest/helpers.go#L95-L131